tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,455 @@
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/kernels/linalg_ops_common.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/kernels/cuda_solvers.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#include "tensorflow/core/util/cuda_kernel_helper.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+// I need to transpose V afterwards
+#include "tensorflow/core/kernels/transpose_functor.h"
+
+// Logging
+#include <stdio.h>
+
+namespace tensorflow {
+
+static const char kErrMsg[] =
+    "Singular Value Decomposition was not successful. The input might not be "
+    "valid.";
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace {
+  template<class Scalar>
+  __global__ void ComputeV1Kernel(CudaLaunchConfig config,
+      int64 m, const Scalar* M, const Scalar* U, const Scalar* S, Scalar* V)
+  {
+    CUDA_1D_KERNEL_LOOP(i, config.virtual_thread_count) {
+      Scalar v = M[i] * U[m*i] * S[0];
+      //printf("i=%d, v=%f\n", (int) i, (float) v);
+      CudaAtomicAdd(V, v);
+    }
+  }
+
+  template<class Scalar>
+  __global__ void ComputeV2Kernel(CudaLaunchConfig config, Scalar* V)
+  {
+    CUDA_1D_KERNEL_LOOP(i, config.virtual_thread_count) {
+      //printf("V'[%d] = %f\n", (int)i, (float)V[i]);
+      V[i] = (Scalar) copysign(1.0, V[i]);
+    }
+  }
+}
+
+// Scalar: The input scalar type (can be complex)
+// SScalar: The output type for the singular value,
+//   same as Scalar if real, or the real version if Scalar is complex
+template <class Scalar, class SScalar>
+class SvdOpGpu : public AsyncOpKernel {
+ public:
+  explicit SvdOpGpu(OpKernelConstruction* context) : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("compute_uv", &compute_uv_));
+    OP_REQUIRES_OK(context, context->GetAttr("full_matrices", &full_matrices_));
+  }
+
+  void RunSVD(OpKernelContext* context, DoneCallback done, int64 m, int64 n,
+              int64 p, int64 batch_size, Scalar* input_ptr,
+              SScalar* outputS_ptr, Scalar* outputU_ptr, Scalar* outputVT_ptr,
+              int* dev_info_ptr, CudaSolver& solver) {
+
+    // Needed for the n=1 fix, see below, since SVD destroys the input
+    Tensor input_copy;
+    if (compute_uv_ && n==1) {
+      OP_REQUIRES_OK_ASYNC(
+        context,
+        context->allocate_temp(DataTypeToEnum<Scalar>::v(), TensorShape({m}), &input_copy),
+        done);
+    }
+
+    for (int64 i = 0; i < batch_size; ++i) {
+      int lda = m;
+      int ldu = m;
+      int ldvt = n;
+      Scalar* input = input_ptr + i * m * n;
+      SScalar* outputS = outputS_ptr + i * p;
+      Scalar* outputU = NULL;
+      Scalar* outputVT = NULL;
+      signed char jobu = 'N';
+      signed char jobvt = 'N';
+
+      // Save input matrix for n=1 fix
+      if (compute_uv_ && n==1) {
```

Comment:
Since the kernel launch overhead is pretty high and the matrix is small in the n==1 case, perhaps you can just create a single temp tensor and copy the entire input outside the batch loop.

Reply:
Yes, I guess the case of a 1000000x1-matrix is pretty rare ;)
But then I could go further and directly call ComputeV1Kernel over multiple batches.

Reply:
My concern is batch_size >> 1, but I suppose we can address it in a followup.